recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch

```diff
@@ -0,0 +1,534 @@
++
++	connector_overlay = malloc(connector_overlay_size);
++	if (!connector_overlay)
++		goto fit_error;
++
++	memcpy(connector_overlay, overlay, connector_overlay_size);
++	return;
++
++fit_error:
++	pr_err("M.2 device tree overlay %s not available,\n", overlay_path);
++#endif
++}
++
++static void m2_connector_setup(void)
++{
++	ulong m2_manual_config = env_get_ulong("m2_manual_config", 10,
++					       CONNECTOR_MODE_INVALID);
++	const char *mode_info = "";
++	struct m2_config_pins config_pins;
++	unsigned int n;
++
++	/* enable M.2 connector power */
++	set_pinvalue("gpio@601000_17", "P3V3_M2_EN", 1);
++	udelay(4 * 100);
++
++	m2_manual_config = env_get_ulong("m2_manual_config", 10,
++					 CONNECTOR_MODE_INVALID);
++	if (m2_manual_config < CONNECTOR_MODE_INVALID) {
++		mode_info = " [manual mode]";
++		connector_mode = m2_manual_config;
++	} else { /* auto detection */
++		for (n = 0; n < ARRAY_SIZE(config_pins.config); n++)
++			config_pins.config[n] =
++				get_pinvalue(m2_bkey_cfg_pin_info[n].gpio_name,
++					     m2_bkey_cfg_pin_info[n].label);
++		connector_mode = CONNECTOR_MODE_INVALID;
++		for (n = 0; n < ARRAY_SIZE(m2_config_table); n++) {
++			if (!memcmp(config_pins.config,
++				    m2_config_table[n].config_pins.config,
++				    sizeof(config_pins.config))) {
++				connector_mode = m2_config_table[n].mode;
++				break;
++			}
++		}
++		if (connector_mode == CONNECTOR_MODE_INVALID) {
++			mode_info = " [fallback, card unknown/unsupported]";
++			connector_mode = BKEY_USB30_EKEY_PCIE;
++		}
++	}
++
++	printf("M.2:   %s%s\n", m2_connector_mode_name[connector_mode],
++	       mode_info);
++
++	/* configure serdes mux */
++	set_pinvalue(serdes_mux_ctl_pin_info[0].gpio_name,
++		     serdes_mux_ctl_pin_info[0].label,
++		     serdes_mux_ctrl[connector_mode].ctrl_usb30_pcie0_lane0);
++	set_pinvalue(serdes_mux_ctl_pin_info[1].gpio_name,
++		     serdes_mux_ctl_pin_info[1].label,
++		     serdes_mux_ctrl[connector_mode].ctrl_pcie1_pcie0);
++	set_pinvalue(serdes_mux_ctl_pin_info[2].gpio_name,
++		     serdes_mux_ctl_pin_info[2].label,
++		     serdes_mux_ctrl[connector_mode].ctrl_usb30_pcie0_lane1);
++
++	m2_overlay_prepare();
++}
++
+ int board_init(void)
+ {
+ 	return 0;
+@@ -215,6 +424,9 @@ int board_late_init(void)
+ 	/* change CTRL_MMR register to let serdes0 not output USB3.0 signals. */
+ 	writel(0x3, SERDES0_LANE_SELECT);
+ 
++	if (board_is_m2())
++		m2_connector_setup();
++
+ 	set_board_info_env();
+ 
+ 	/* remove the eMMC if requested via button */
+@@ -226,6 +438,29 @@ int board_late_init(void)
+ }
+ 
+ #if defined(CONFIG_OF_LIBFDT) && defined(CONFIG_OF_BOARD_SETUP)
++static void m2_fdt_fixup(void *blob)
++{
++	void *overlay;
++	int err;
++
++	if (!connector_overlay)
++		return;
++
++	overlay = malloc(connector_overlay_size);
++	if (!overlay)
++		goto fixup_error;
++
++	memcpy(overlay, connector_overlay, connector_overlay_size);
++	err = fdt_overlay_apply_verbose(blob, overlay);
++	free(overlay);
```

Comment:
two malloc one free

Reply:
The first malloc is (almost) persistent: we only free the first copy implicitly when U-Boot boot-time services terminate (on hand-over to OS during legacy boot or when UEFI boot-time services are terminated). Thus no free.

Reply:
Then we hard-code in the code the max size of the overlay.